qlib/rl/order_execution/reward.py

```diff
@@ -47,3 +48,51 @@ def reward(self, simulator_state: SAOEState) -> float:
         self.log("reward/pa", pa)
         self.log("reward/penalty", penalty)
         return reward * self.scale
+
+
+def _weighted_average(val: np.ndarray, weight: np.ndarray):
+    return (val * weight).sum() / weight.sum()
+
+
+class PPOReward(Reward[SAOEState]):
+    """Reward proposed by paper "An End-to-End Optimal Trade Execution Framework based on Proximal Policy Optimization".
+    
+    Parameters
+    ----------
+    max_step
+        Maximum number of steps.
+    start_time_index
+        First time index that allowed to trade.
+    end_time_index
+        Last time index that allowed to trade.
+    """
+    
+    def __init__(self, max_step: int, start_time_index: int = 0, end_time_index: int = 239) -> None:
+        self.max_step = max_step
+        self.start_time_index = start_time_index
+        self.end_time_index = end_time_index
+    
+    def reward(self, simulator_state: SAOEState) -> float:
+        if simulator_state.cur_step == self.max_step - 1 or simulator_state.position < 1e-6:
+            traded = simulator_state.order.amount - simulator_state.position
+            trade_value = sum(simulator_state.history_steps["trade_value"])
+            intraday_data = simulator_state.backtest_data.data[self.start_time_index:self.end_time_index + 1]
+            
+            if traded >= 1e-6:
+                vwap_price = trade_value / traded
+            else:
+                vwap_price = _weighted_average(  # VWAP price of the entire day
+                    val=intraday_data["$close0"].to_numpy(), 
+                    weight=intraday_data["$volume0"].to_numpy(),
+                )
+            twap_price = np.mean(intraday_data["$close0"].to_numpy())
```

Comment:
We should use actual deal price instead of $close0. Might use
```
vwap_price = simulator_state.metrics.trade_price
twap_price = simulator_state.backtest_data.get_deal_price().mean()
```

Reply:
Also, we need to add the judgement for twap = 0.